operator/duties/scheduler.go

```diff
@@ -0,0 +1,344 @@
+		s.currentDutyDependentRoot = data.CurrentDutyDependentRoot
+
+		currentTime := time.Now()
+		delay := s.network.SlotDurationSec() / 3 /* a third of the slot duration */
+		slotStartTimeWithDelay := s.slotStartTime(data.Slot).Add(delay)
+		if currentTime.Before(slotStartTimeWithDelay) {
+			logger.Debug("🏁 Head event: Block arrived before 1/3 slot", zap.Duration("time_saved", slotStartTimeWithDelay.Sub(currentTime)))
+			s.waitMutex.Lock()
+			s.waitCond.Broadcast()
+			s.waitMutex.Unlock()
+		}
+	}
+}
+
+// ExecuteDuties tries to execute the given duties
+func (s *Scheduler) ExecuteDuties(logger *zap.Logger, duties []*spectypes.Duty) {
+	for _, duty := range duties {
+		loggerWithContext := s.loggerWithDutyContext(logger, duty)
+
+		dutyToExecute := duty
+		go func() {
+			err := s.ExecuteDuty(loggerWithContext, dutyToExecute)
+			if err != nil {
+				loggerWithContext.Error("failed to execute duty", zap.Error(err))
+			}
+		}()
+	}
+}
+
+func (s *Scheduler) ExecuteDuty(logger *zap.Logger, duty *spectypes.Duty) error {
+	if s.executor != nil {
+		// enables to work with a custom executor, e.g. readOnlyDutyExec
+		return s.executor.ExecuteDuty(logger, duty)
+	}
+
+	if duty.Type == spectypes.BNRoleAttester || duty.Type == spectypes.BNRoleSyncCommittee {
+		s.waitOneThirdOrValidBlock(duty.Slot)
+	}
+
+	// because we're using the same duty for more than 1 duty (e.g. attest + aggregator) there is an error in bls.Deserialize func for cgo pointer to pointer.
+	// so we need to copy the pubkey val to avoid pointer
+	var pk phase0.BLSPubKey
+	copy(pk[:], duty.PubKey[:])
+
+	if v, ok := s.validatorController.GetValidator(hex.EncodeToString(pk[:])); ok {
+		ssvMsg, err := CreateDutyExecuteMsg(duty, pk, s.network.Domain)
+		if err != nil {
+			return err
+		}
+		dec, err := queue.DecodeSSVMessage(logger, ssvMsg)
+		if err != nil {
+			return err
+		}
+		if pushed := v.Queues[duty.Type].Q.TryPush(dec); !pushed {
+			logger.Warn("dropping ExecuteDuty message because the queue is full")
+		}
+		// logger.Debug("📬 queue: pushed message", fields.MessageID(dec.MsgID), fields.MessageType(dec.MsgType))
+	} else {
+		logger.Warn("could not find validator")
+	}
+
+	return nil
+}
+
+// loggerWithDutyContext returns an instance of logger with the given duty's information
+func (s *Scheduler) loggerWithDutyContext(logger *zap.Logger, duty *spectypes.Duty) *zap.Logger {
+	return logger.
+		With(fields.Role(duty.Type)).
+		With(zap.Uint64("committee_index", uint64(duty.CommitteeIndex))).
+		With(fields.CurrentSlot(s.network.Beacon)).
+		With(fields.Slot(duty.Slot)).
+		With(fields.Epoch(s.network.Beacon.EstimatedEpochAtSlot(duty.Slot))).
+		With(fields.PubKey(duty.PubKey[:])).
+		With(fields.StartTimeUnixMilli(s.network.Beacon, duty.Slot))
+}
+
+// waitOneThirdOrValidBlock waits until one-third of the slot has transpired (SECONDS_PER_SLOT / 3 seconds after the start of slot)
+func (s *Scheduler) waitOneThirdOrValidBlock(slot phase0.Slot) {
+	s.waitMutex.Lock()
```

Comment:
Can we have a test for this method to make sure that locking `waitMutex` here and in the goroutine doesn't lead to a deadlock?

Reply:
the unit tests covering it